File: src/main/java/io/trydent/olimpo/io/Port.java
package io.trydent.olimpo.io;

import io.trydent.olimpo.sys.Property;
import io.trydent.olimpo.type.Type;
import org.slf4j.Logger;

import java.util.Optional;

import static io.trydent.olimpo.sys.Property.envVar;
import static org.slf4j.LoggerFactory.getLogger;

public interface Port extends Type.AsInt {
  static Port envPort(String name, int orDefault) {
    return new EnvPort(envVar(name), portOrDie(orDefault));
  }

  static Optional<Port> port(int value) {
    return Optional.of(value)
      .filter(it -> it > 0)
      .map(PortImpl::new);
  }

  static Port portOrDie(int port) {
    return port(port).orElseThrow(IllegalArgumentException::new);
  }
}

final class PortImpl implements Port {
  private final int port;

  PortImpl(final int port) {
    this.port = port;
  }

  @Override
  public final int get() {
    return port;
  }
}

final class EnvPort implements Port {
  private static final Logger log = getLogger(EnvPort.class);

  private final Property property;
  private final Port orDefault;

  EnvPort(final Property property, final Port orDefault) {
    this.property = property;
    this.orDefault = orDefault;
  }

  @Override
  public final int get() {
    try {
      return Integer.parseInt(property.get());
    } catch (NumberFormatException nfe) {
      log.warn("Provided port is not a number.");
      return orDefault.get();
    }
  }
}

File: src/main/java/io/trydent/olimpo/io/Port.kt
package io.trydent.olimpo.io

import io.trydent.olimpo.sys.Property
import io.trydent.olimpo.sys.Property.Companion.envVar

@Suppress("MemberVisibilityCanBePrivate")
interface Port : () -> Int {
  companion object {
    fun envPort(name: String, default: Int): Port = EnvPort(envVar(name), port(default))

    fun port(value: Int): Port? = value
      .takeIf { it > 0 }
      ?.let(::PortImpl)
  }
}

internal class PortImpl(private val port: Int) : Port {
  override fun invoke(): Int = port

  override fun toString() = "${this()}"
}

internal class EnvPort(private val property: Property, private val default: Port?) : Port {
  override fun invoke(): Int = try {
    Integer.parseInt(property())
  } catch (nfe: NumberFormatException) {
    default!!()
  }

  override fun toString() = "${this()}"
}

